fuse_constraints/include/fuse_constraints/normal_prior_pose_3d_cost_functor.h

```diff
@@ -0,0 +1,127 @@
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_CONSTRAINTS_NORMAL_PRIOR_POSE_3D_COST_FUNCTOR_H
+#define FUSE_CONSTRAINTS_NORMAL_PRIOR_POSE_3D_COST_FUNCTOR_H
+
+#include <fuse_constraints/util.h>
+#include <fuse_constraints/normal_prior_orientation_3d_cost_functor.h>
+
+#include <ceres/internal/disable_warnings.h>
+#include <ceres/internal/eigen.h>
+#include <Eigen/Core>
+
+
+namespace fuse_constraints
+{
+
+/**
+ * @brief Create a prior cost function on both the 3D position and orientation variables at once.
+ *
+ * The Ceres::NormalPrior cost function only supports a single variable. This is a convenience cost function that
+ * applies a prior constraint on both the 3D position and orientation variables at once.
+ *
+ * The cost function is of the form:
+ *
+ *             ||    [  x       -     b(0)  ] ||^2
+ *             ||    [  y       -     b(1)  ] ||
+ *             ||    [  z       -     b(2)  ] ||
+ *             ||A * [  (b(4:7) * q^-1)(1)  ] ||
+ *   cost(x) = ||    [  (b(4:7) * q^-1)(2)  ] ||
+ *             ||    [  (b(4:7) * q^-1)(3)  ] ||
+ *
+ * where, the matrix A and the vector b are fixed and (x, y, z, qw, qx, qy, qz) are the components of the position and
+ * orientation variables. Note that the covariance submatrix for the quaternion is 3x3, representing errors around each
+ * imaginary component axis. In case the user is interested in implementing a cost function of the form
+ *
+ *   cost(X) = (X - mu)^T S^{-1} (X - mu)
+ *
+ * where, mu is a vector and S is a covariance matrix, then, A = S^{-1/2}, i.e the matrix A is the square root
+ * information matrix (the inverse of the covariance).
+ */
+class NormalPriorPose3DCostFunctor
+{
+public:
+  using Matrix6d = Eigen::Matrix<double, 6, 6>;
+  using Vector7d = Eigen::Matrix<double, 7, 1>;
+
+  /**
+   * @brief Construct a cost function instance
+   *
+   * @param[in] A The residual weighting matrix, most likely the square root information matrix in order
+   *              (x, y, z, qw, qx, qy, qz)
+   * @param[in] b The 3D pose measurement or prior in order (x, y, z, qw, qx, qy, qz)
+   */
+  NormalPriorPose3DCostFunctor(const Matrix6d& A, const Vector7d& b);
+
+  /**
+   * @brief Evaluate the cost function. Used by the Ceres optimization engine.
+   */
+  template <typename T>
+  bool operator()(const T* const position, const T* const orientation, T* residual) const;
+
+private:
+  Matrix6d A_;
+  Vector7d b_;
+
+  NormalPriorOrientation3DCostFunctor orientation_functor_;
+};
+
+NormalPriorPose3DCostFunctor::NormalPriorPose3DCostFunctor(const Matrix6d& A, const Vector7d& b) :
+  A_(A),
+  b_(b),
+  orientation_functor_(Eigen::Matrix3d::Identity(), b_.tail<4>())  // Delta will not be scaled
```

Comment:
So, we have two choices here:

1. Pass in the submatrix of `A_` that contains the covariance for the orientation here. Then, in `operator()`, we can only apply the covariance to the `position` variables.

2. Pass in the identity here so that whatever scaling is done has no effect, and then multiply the full `A_` by the entire residual.

I opted for (2), but could be convinced to do (1).

Reply:
(2) is definitely the better answer.

If we ever get so concerned about performance that the needless scaling by identity is an issue, I would reimplement the quaternion cost here.